# Credit Risk Classification

Credit risk poses a classification problem that’s inherently imbalanced. This is because healthy loans easily outnumber risky loans. In this Challenge, you’ll use various techniques to train and evaluate models with imbalanced classes. You’ll use a dataset of historical lending activity from a peer-to-peer lending services company to build a model that can identify the creditworthiness of borrowers.

## Instructions:

This challenge consists of the following subsections:

* Split the Data into Training and Testing Sets

* Create a Logistic Regression Model with the Original Data

* Predict a Logistic Regression Model with Resampled Training Data 

### Split the Data into Training and Testing Sets

Open the starter code notebook and then use it to complete the following steps.

1. Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

2. Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

    > **Note** A value of `0` in the “loan_status” column means that the loan is healthy. A value of `1` means that the loan has a high risk of defaulting.  

3. Check the balance of the labels variable (`y`) by using the `value_counts` function.

4. Split the data into training and testing datasets by using `train_test_split`.

### Create a Logistic Regression Model with the Original Data

Employ your knowledge of logistic regression to complete the following steps:

1. Fit a logistic regression model by using the training data (`X_train` and `y_train`).

2. Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

3. Evaluate the model’s performance by doing the following:

    * Calculate the accuracy score of the model.

    * Generate a confusion matrix.

    * Print the classification report.

4. Answer the following question: How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

### Predict a Logistic Regression Model with Resampled Training Data

Did you notice the small number of high-risk loan labels? Perhaps, a model that uses resampled data will perform better. You’ll thus resample the training data and then reevaluate the model. Specifically, you’ll use `RandomOverSampler`.

To do so, complete the following steps:

1. Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

2. Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

3. Evaluate the model’s performance by doing the following:

    * Calculate the accuracy score of the model.

    * Generate a confusion matrix.

    * Print the classification report.
    
4. Answer the following question: How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

### Write a Credit Risk Analysis Report

For this section, you’ll write a brief report that includes a summary and an analysis of the performance of both machine learning models that you used in this challenge. You should write this report as the `README.md` file included in your GitHub repository.

Structure your report by using the report template that `Starter_Code.zip` includes, and make sure that it contains the following:

1. An overview of the analysis: Explain the purpose of this analysis.


2. The results: Using bulleted lists, describe the balanced accuracy scores and the precision and recall scores of both machine learning models.

3. A summary: Summarize the results from the machine learning models. Compare the two versions of the dataset predictions. Include your recommendation for the model to use, if any, on the original vs. the resampled data. If you don’t recommend either model, justify your reasoning.

In [36]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from imblearn.metrics import classification_report_imbalanced

import warnings
warnings.filterwarnings('ignore')

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [37]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
lending_df = pd.read_csv(
    Path("./Resources/lending_data.csv")
)

# Review the DataFrame
display(lending_df.head())
display(lending_df.tail())

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

loan_size  interest_rate  borrower_income  debt_to_income  \
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  loan_status  
77531               12                 2       56600            1  
77532               11                 2       50900            1  
77533               11                 2       50300            1  
77534               10                 2       45300            1  
77535                9                 2       42300            1

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [38]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = lending_df["loan_status"]

# Separate the X variable, the features
X = lending_df.drop(columns=["loan_status"])

In [39]:
# Review the y variable Series
print(y)

0        0
1        0
2        0
3        0
4        0
        ..
77531    1
77532    1
77533    1
77534    1
77535    1
Name: loan_status, Length: 77536, dtype: int64


In [40]:
# Review the X variable DataFrame
display(X.head())
display(X.tail())

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

loan_size  interest_rate  borrower_income  debt_to_income  \
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  
77531               12                 2       56600  
77532               11                 2       50900  
77533               11                 2       50300  
77534               10                 2       45300  
77535                9                 2       42300

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [41]:
# Check the balance of our target values
y.value_counts()

0    75036
1     2500
Name: loan_status, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [42]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [43]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(random_state=1)

# Fit the model using training data
lr_model = logistic_regression_model.fit(X_train, y_train)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [46]:
# Make a prediction using the testing data
y_original_prediction = lr_model.predict(X_test)
y_original_prediction

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [48]:
# Print the balanced_accuracy score of the model
balanced_accuracy_score(y_test, y_original_prediction)

0.9520479254722232

In [49]:
# Generate a confusion matrix for the model
confusion_matrix(y_test, y_original_prediction)

array([[18663,   102],
       [   56,   563]], dtype=int64)

In [50]:
# Print the classification report for the model
print(classification_report_imbalanced(y_test, y_original_prediction))

                   pre       rec       spe        f1       geo       iba       sup

          0       1.00      0.99      0.91      1.00      0.95      0.91     18765
          1       0.85      0.91      0.99      0.88      0.95      0.90       619

avg / total       0.99      0.99      0.91      0.99      0.95      0.91     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** 

**Findings:**
Following are the findings for the Classification Report of Original Prediction with Target Test (y_test) attribute: 


**Accuracy Score:** 
The Accuracy Score after considering TPs, FPs, TNs & FNs, we get the pretty high score of 0.95 i.e. 95% of accuracy in the performance of this Model. Therefore, the overall evaluation of this model states the accuracy level of 95% without breaking down at Risky & Non-Risky Loans levels.

Accuracy Score just calculates & signifies the overall confidence level of Model, and in this case when we run the Classification Report for Original Prediction with Target Test variable (y_test) it just shows that 95% accuracy rate of this Model nothing more.


**Precision Score:** 
The Precision Score is more precise than Accuracy Score for obvious reasons which is that it doesn't consider the Negatives and only focus on Positives (TPs & FPs) & particularly in this finding after considering TPs & FPs, we get:

FPs as High-Risk Loans (Value 1) was 0.85 (85%) with less confidence compared to its counterpart TPs as Healthy Loans with the reading of 1.00 (100%) i.e. the highest level of confidence in Model's Performance. Hence, the Precision Score determines the model's confidence in Healthy Loans is more as compare to Risky Loans. 


**Recall Score:**
The purpose of Recall Score is to measure the True Judgements of Model, meaning we human knows that how many are True Positive Results & False Negative Results in total which the model has missed. 

Therefore, after considering TPs & FNs we get:
For Value 1, 0.91 (91%) for High-Risk Loans shows the Recall Score is quite high. Whereas for Value 0, 0.99 (99%) for Healthy Loans which is quite similar to Precision Score (100%).


**F1 Score:**
F1 Score is basically a Harmonic Mean, as a single summary statistic for the Precision and the Recall. Which actually computes after getting the Precision & Recall values and applying those values in this formula: 2 × (precision x recall) / (precision + recall). 

Therefore, after computing F1 Score we get:
For Value 1, 0.88 (88%) for High-Risk Loans shows the F1 Score is actually somewhere in the middle of Precision & Recall Values. Where, the Precision was 85% & Recall was 91% we get the more refined value i.e. 88% after applying the F1 Score formula. Whereas for Value 0, 1.00 (100%) confidence at 3rd level of calculation after computing Precision, Recall & F1 proves that how solid the confifence is actually lies in Healthy Loans.


**Conclusion:**
The Model Performance has been measured or evaluated by applying confusion matrix in general where we have seen that the confidence for the Values 1 & 0 with following conclusion for each of them:

**1. Value 1 (High-Risk Loans):**
The Value 1 in particular has been varied in different accuracy tests and we get the range of 85-91% (Precision & Recall) after applying the Precision & Recall techniques. Whereas, the F1 Score which is a Harmonic Value for the Risky Loans shows the best possible value lies witin the range i.e. 88%. 

This 88% shows the best possible value as the accuracy of this model after considering the TPs, FPs & FNs determining that how accurate the performance of this model is in reality.

**2. Value 0 (Healthy Loans):**
The Value 0 shows the highest level of confidence in all 3 techniques: Precision: 100%, Recall: 99% & F1 Score: 100%. This shows that the confidence in measuring the Larger Class of Value 0 (Healthy Loans) is way too high. Machine may sound pretty confident at this level however this might get wrong in long run where machine model develops a tendency to favor Value 0 and give more precedence over Value 1.

Therefore, to avoid the unbiased behaviour we have a functionality called: Imbalanced Classes where we apply Resampling Techniques. Specifically, Oversampling is one the major technique will be leveraged for this purpose to artificially increase the size of Smaller Class so that the comparison with the Larger Class will be trated equal without getting biased, hence to get more fair results by eliminating error at fractional level.       

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [13]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
random_oversampler_model = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
x_resampled, y_resampled = random_oversampler_model.fit_resample(X_train, y_train)

In [14]:
# Count the distinct values of the resampled labels data
y_resampled.value_counts()

0    56271
1    56271
Name: loan_status, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [22]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(random_state=1)

# Fit the model using the resampled training data
lr_resampled_model = logistic_regression_model.fit(x_resampled, y_resampled)

# Make a prediction using the testing data
y_resampled_pred = lr_resampled_model.predict(X_test)
y_resampled_pred

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [23]:
# Print the balanced_accuracy score of the model 
balanced_accuracy_score(y_test, y_resampled_pred)

0.9936781215845847

In [25]:
# Generate a confusion matrix for the model
confusion_matrix(y_test, y_resampled_pred)

array([[18649,   116],
       [    4,   615]], dtype=int64)

In [27]:
# Print the classification report for the model
print(classification_report_imbalanced(y_test, y_resampled_pred))

                   pre       rec       spe        f1       geo       iba       sup

          0       1.00      0.99      0.99      1.00      0.99      0.99     18765
          1       0.84      0.99      0.99      0.91      0.99      0.99       619

avg / total       0.99      0.99      0.99      0.99      0.99      0.99     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** 

**Findings:**

After applying: **RandomOverSampler Technique**
Following are the findings for the Imbalanced Classification Report of Resampled Data (Predictions) with Target Test (y_test) attribute: 


**Accuracy Score:** 
We get the Accuracy Score after resampling is 0.99 (99%) as compare to the Original Data's Accuracy Score which was 95%, it shows that after applying OverSampling Technique it actually improves the accuracy level.

Hence, the imbalanced oversampling technique improves the confidence level of model.  



**Precision Score:** 
We get the following Precision Score readings for Values: 1 & 0 after Resampling:

1. FPs as High-Risk Loans (Value 1): Precision Score is 0.84 (84%) as compare to Original Data's Precision Score which was 85%, it's very slightly less than the original ones however this 1% change won't make big difference therefore it must be considered as same without much change.

2. TPs as Healthy Loans (Value 0): Precision Score is 1.00 (100%) as compare to Original Data's Precision Score which was 100%, therefore no change in Precision Score for Healthy Loans. 



**Recall Score:**
We get the following Recall Score readings for Values: 1 & 0 after Resampling:

1. TPs as Healthy Loans (Value 0): Recall Score is 0.99 (99%) as compare to Original Data's Recall Score which was 99% too, therefore there's no change in Recall Score for Healthy Loans.

2. FNs as High-Risk Loans (Value 1): Recall Score is 0.99 (99%) as compare to Original Data's Recall Score which was 91%, this proves that how Imbalaced OverSampling can be useful. There's an 8% increase can be seen in confidence of this model after applying Resampling Techniques.



**F1 Score:**
We get the following F1 Score readings for Values: 1 & 0 after Resampling:

1. High-Risk Loans (Value 1): F1 Score is 0.91 (91%) as compare to Original Data's F1 Score which was 88%, again proving that Imbalanced OverSampling Technique is quite useful in making more rational decisions. There's a 3% hike can be seen in confidence of this model after applying Resampling Techniques.

2. Healthy Loans (Value 0): F1 Score is 1.00 (100%) as compare to Original Data's F1 Score which was 100%, hence no change in confidence. It always been on seventh cloud when it comes to Healthy Loans, apparently Model loves Healthy Loan Data. 



**Conclusion:**
After applying Resampling Technique, we have seen some changes particularly for Smaller Class (High-Risk Loans) in case of:

1. Precision Score: Where Original Data's Precision Score was 85% and Resampled Data's Precision Score was 84%, which is quite same not much difference.

2. F1 Score: Where Original Data's F1 Score was 88% and Resampled Data's F1 Score was 91%, an increase in confidence level by 3% is quite impressive and shows that how Harmonic Mean value has been increased in total making the model efficacy more efficient and accurate.

Value 1 (Healthy Loans) remains pretty consistent in our findings and nothing major changed at confidence level and looks like all the readings regardless the Original or Resampled data, the nature is quite similar in both. 